### PR TITLE
wordpress: replace the dbPassword option with dbPasswordFile

### PR DESCRIPTION
**Pull request by basvandijk**
We shouldn't force users to store passwords in the world-readable Nix store so this commit replaces the `dbPassword` option with `dbPasswordFile` which should be set to a file that contains the wordpress DB password. Users can then provision this file however they want, either by storing it unsafely in the Nix store using `pkgs.writeTextFile` or safely using something like [nixops key management](https://nixos.org/nixops/manual/#idm140737318306400). 

It would be great if this can also be cherry-picked on `release-17.03`.

### CODE REVIEW

**Comment by globin** [comment]
I'd prefer to have both options.

**Comment by basvandijk** [comment]
@globin I prefer to make users conscious about the safety of keys. An option like `dbPassword` kind of hides the fact that the key is stored in the world-readable Nix store. Of course documentation can help but not everybody reads that. 

The `dbPasswordFile` option forces users to think where they are going to store the key. 

Of course the old *unsafe* behaviour is easy to recreate using `pkgs.writeTextFile` as I [do in the test](https://github.com/NixOS/nixpkgs/pull/24146/files#diff-4f2d880218c011a3aa8ce640e167be5aR15).

**Comment by basvandijk** [comment]
@globin I changed my mind and followed your preference. `dbPasswordFile` now defaults to a file in the Nix store with the value of `dbPassword`. I added documentation to warn users about the security issues.

The nice thing is that this is also backwards compatible.

**Comment by LnL7** [comment]
It's possible to add [`warnings`](https://search.nix.gsc.io/?q=warnings%20%3D&i=nope&files=&repos=) when using certain options.

**Review by fpletz — APPROVED** [review]

**Comment by RonnyPfannschmidt** [comment]
btw, a is, this one is a breaking change

**Comment by basvandijk** [comment]
@RonnyPfannschmidt what breaks because of this change?

**Comment by RonnyPfannschmidt** [comment]
@basvandijk any user that has this already configured afterwards has a broken configuration, because the old thing is gone, there is not the slightest hint of a transition help either

**Comment by basvandijk** [comment]
@RonnyPfannschmidt the old thing isn't gone. The change should be backwards compatible.

**Comment by RonnyPfannschmidt** [comment]
indeed, i am sorry, i missed the `pkgs.writeTextFile` bit

File: nixos/modules/services/web-servers/apache-httpd/wordpress.nix [hunk 1]
```diff
@@ -137,9 +137,33 @@ in
...
+
+        Warning: this password is stored in the world-readable Nix store. It's
+        recommended to use the $dbPasswordFile option since that gives you control over
+        the security of the password. $dbPasswordFile also takes precedence over $dbPassword.
+      '';
       example = "wordpress";
     };
+    dbPasswordFile = mkOption {
+      type = types.str;
+      default = toString (pkgs.writeTextFile {
```
Comment (LnL7) [inline nixos/modules/services/web-servers/apache-httpd/wordpress.nix]: I'm not sure, but I think more complex defaults like this are usually moved to the implementation with a `mkDefault`

**Review by LnL7 — APPROVED** [review]

File: nixos/modules/services/web-servers/apache-httpd/wordpress.nix (hunk 1 above)
Comment (globin) [inline nixos/modules/services/web-servers/apache-httpd/wordpress.nix]: Correct, otherwise the manual is rebuilt when `dbPassword` changes in your config.

Comment (LnL7) [inline nixos/modules/services/web-servers/apache-httpd/wordpress.nix]: Also, where's the config attribute for this module?

Comment (basvandijk) [inline nixos/modules/services/web-servers/apache-httpd/wordpress.nix]: @globin note that wordpress is not mentioned in the manual.

@LnL7 `wordpress.nix` is not a traditional NixOS module and doesn't have a `config` attribute. It should be used as:

```nix
{ 
  services.httpd.extraSubservices = [
    { serviceType = "wordpress";
      dbName = "...";
      themes = [];
      plugins = [];
    }
  ];
}
```

So there's no need and no place to use `mkDefault`.

Comment (LnL7) [inline nixos/modules/services/web-servers/apache-httpd/wordpress.nix]: I see 😕. https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/apache-httpd/per-server-options.nix#L145

**Comment by basvandijk** [comment]
Similarly to [other web-apps](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/pump.io.nix#L282) I added an example to the `dbPasswordFile` option: `"/run/keys/wordpress-dbpassword"`.

Is there anything else that prevents this being mergeable?

**Comment by qknight** [comment]
love this concept, we should not stop at password though. `wordpress` also contains other key materials as `AUTH_KEY`, `AUTH_SALT`, `...` shown here:

     https://api.wordpress.org/secret-key/1.1/salt/

same is true for owncloud/nextcloud and so on.

**Comment by basvandijk** [comment]
@qknight thanks and I agree that `AUTH_KEY` and so forth should not be placed in the Nix store as well. However at the moment the wordpress module doesn't set these parameters so (apart from the `dbPassword`) users aren't currently being forced to do something unsafe.

**Comment by basvandijk** [comment]
@qknight thanks for the merge! Can this also be cherry-picked on `release-17.03`? I would like to use this in production.

**Comment by basvandijk** [comment]
Hi @qknight, it would be great if you can cherry pick this on `release-17.03`.

**Comment by qknight** [comment]
@basvandijk added it to `release-17.03` just now. please check if it works for you.

**Pull request merged**